source/adios2/core/IO.cpp

```diff
@@ -0,0 +1,279 @@
+/*
+ * Distributed under the OSI-approved Apache License, Version 2.0.  See
+ * accompanying file Copyright.txt for details.
+ *
+ * IO.cpp
+ *
+ *  Created on: Jan 6, 2017
+ *      Author: William F Godoy godoywf@ornl.gov
+ */
+
+#include "IO.h"
+#include "IO.tcc"
+
+#include "adios2/ADIOSMPI.h"
+#include "adios2/helper/adiosFunctions.h" //BuildParametersMap
+
+#include "adios2/engine/bp/BPFileWriter.h"
+
+#ifdef ADIOS2_HAVE_DATAMAN // external dependencies
+#include "adios2/engine/dataman/DataManReader.h"
+#include "adios2/engine/dataman/DataManWriter.h"
+#endif
+
+#ifdef ADIOS2_HAVE_ADIOS1 // external dependencies
+#include "adios2/engine/adios1/ADIOS1Reader.h"
+#include "adios2/engine/adios1/ADIOS1Writer.h"
+#endif
+
+#ifdef ADIOS2_HAVE_HDF5 // external dependencies
+#include "adios2/engine/hdf5/HDF5ReaderP.h"
+#include "adios2/engine/hdf5/HDF5WriterP.h"
+#endif
+
+namespace adios
+{
+
+IO::IO(const std::string name, MPI_Comm mpiComm, const bool inConfigFile,
+       const bool debugMode)
+: m_Name(name), m_MPIComm(mpiComm), m_InConfigFile(inConfigFile),
+  m_DebugMode(debugMode)
+{
+}
+
+void IO::SetEngine(const std::string engineType) { m_EngineType = engineType; }
+void IO::SetIOMode(const IOMode ioMode) { m_IOMode = ioMode; };
+
+VariableCompound &IO::GetVariableCompound(const std::string &name)
```

Comment:
Those returning an internal reference can't be const. I will change the others, though.